cilium/CFP-39876-clustermesh-filtered-export.md

```diff
@@ -0,0 +1,110 @@
+The cache maps backend IP addresses to metadata including service name and namespace
+
+#### Cache Management
+
+The cache is populated via service endpoint updates and receives live updates when backends change. Stale entries are automatically cleaned up, and synchronization mechanisms ensure the global service endpoint cache contains the latest copy of backend endpoint ips. Both Cilium Agent and Cilium Operator already listen to existing Service Events. These events are consumed to populate the global service backend IP cache. 
+
+### Cilium Agent Changes
+
+#### Service Cache Extension
+
+The agent's existing service cache is extended to support lookup for ips against global services.
+
+#### CiliumEndpoint Reconciliation
+
+Every 10 seconds, the agent checks each CiliumEndpoint IP against the global backend cache. If matched, it adds an internal annotation to mark it as a global backend; otherwise, it removes the annotation. This annotation is internal and not user-modifiable.
+
+#### CiliumIdentity Reconciliation
+
+When managing CiliumIdentities, the agent listens for global service IP cache events. On upserts or deletions, it reconciles the relevant identities and updates their annotations accordingly.
+
+### Cilium Operator Changes
+
+#### Service Cache Extension
+
+The agent's existing service cache is extended to support lookup for ips against global services.
+
+#### CiliumIdentity Handling
+
+If the operator manages CiliumIdentities, it responds to global service events from its cache and performs reconciliation to add or remove internal annotation as needed.
+
+#### CiliumEndpointSlice Controller Updates
+
+The controller, which already watches CiliumEndpoint resources, is updated to check for the global backend annotation. If present, it adds a corresponding annotation to the CiliumEndpointSlice. Annotation changes are tracked and updated accordingly.
+
+### ClusterMesh API Server Changes
+
+#### Export Filtering
+
+The ClusterMesh API server filters resources before exporting to etcd. Only those with the internal global backend annotation are shared for cross-cluster visibility.
+
+#### Supported Resource Types
+
+The filter applies to CiliumEndpoint, CiliumIdentity, and CiliumEndpointSlice. Only annotated resources are exported to be shared cross clusters.
+
+## Future Milestones
+
+### Dynamic Export Scope Configuration
+
+Introduce support for dynamic configuration of export scope per namespace or pod label. This would allow operators to fine-tune which endpoints and identities are shared across clusters beyond just global service association.
+
+### Policy-Aware Export Filtering
+
+Integrate network policy awareness into the export decision logic. Endpoints and identities could be exported based on whether they are referenced in cross-cluster network policies, even if not fronted by a global service.
```

Comment:
> Integrate network policy awareness into the export decision logic.

This is not necessarily possible, unless policy information is additionally exchanged cross-cluster. Consider for instance the following scenario:

* Client A1 hosted in cluster A
* Server B1 hosted in cluster B
* A1 wants to access B1 via a global service
* B1 is matched by an ingress policy to allow incoming traffic from A1

To make that work, cluster B needs to be aware of A1's IP to identity mapping (i.e., its CiliumEndpoint), at least in native routing mode, and A1's identity details (i.e., the CiliumIdentity). However, cluster A would have no ways of knowing that such a policy exists in B, and hence that these resources would need to be shared.

Reply:
It should be integrated in this PR as well, as it serves as future reference.